### PR TITLE
Introduce TaskValue to avoid out-of-graph execution

### PR DESCRIPTION
**Pull request by dwijnand**
.. also upgrade to sbt 1.1.0 & cleanup.

### CODE REVIEW

File: src/sbt-test/sbt-buildinfo/task/build.sbt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+val projInTaskGraph1 = project settings (
+  sourceGenerators in Compile += Def.task { counterInTaskGraph.incrementAndGet(); Nil }.taskValue
+)
+
+val projInTaskGraph2 = project dependsOn projInTaskGraph1 settings (
+  BuildInfoPlugin.buildInfoDefaultSettings,
+  addBuildInfoToConfig(Test),
+  buildInfoKeys in Test += (fullClasspath in Compile).taskValue
```
Comment (eed3si9n) [inline src/sbt-test/sbt-buildinfo/task/build.sbt]: In https://github.com/sbt/sbt/pull/2943 I'm injecting `.taskValue` as macro when the LHS is `Initialize[Seq[Task[...]]]`.
Would you be able to take advantage of that if `buildInfoKeys` were `SettingKey[Seq[Task[String]]`, and you unify other miscellaneous build info entries to a `Task[...]` instance?

Comment (dwijnand) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: I'm not sure - maybe. But how are you going to change the type of `buildInfoKeys`?

Comment (eed3si9n) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: We can just change the type and bump up minor version.

Comment (dwijnand) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: What about any plugin that depends on sbt-buildinfo?

Comment (eed3si9n) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: I am personally unaware of any such plugin, and I don't feel obligation to keep bincompat as long as we bump to 0.8.0.

Comment (dwijnand) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: I don't think this is going to work because then this no longer works:

Comment (dwijnand) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: or (and this was the same argument as sbt/sbt#2943) own the fact that `TaskKey`s and `Task` values are different types in sbt, and there are different situations for using either.

Comment (dwijnand) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: Perhaps:

Comment (dwijnand) [reply src/sbt-test/sbt-buildinfo/task/build.sbt]: Perhaps:

**Comment by dwijnand** [comment]
@eed3si9n have a look, and let me know what you think now

File: src/main/scala/sbtbuildinfo/package.scala [hunk 2]
```diff
@@ -4,23 +4,34 @@ package object sbtbuildinfo {
...
     def apply[A](key: TaskKey[A]): Entry[A] = Task(key)
     def apply[A: Manifest](tuple: (String, A)): Entry[A] = Constant(tuple)
     def map[A, B: Manifest](from: Entry[A])(fun: ((String, A)) => (String, B)): Entry[B] =
       BuildInfoKey.Mapped(from, fun)
     def action[A: Manifest](name: String)(fun: => A): Entry[A] = Action(name, () => fun)
 
+    def of(x: Any): BuildInfoKey = macro BuildInfoKeyMacros.ofImpl
+    def ofN(xs: Any*): Seq[BuildInfoKey] = macro BuildInfoKeyMacros.ofNImpl
+
+    def outOfGraphTaskExecution[A](key: TaskKey[A]): Entry[A] = Task(key)
```
Comment (eed3si9n) [inline src/main/scala/sbtbuildinfo/package.scala]: Maybe shorter name like `outOfGraph(..)`? Also is this tested ?

Comment (dwijnand) [reply src/main/scala/sbtbuildinfo/package.scala]: I went with the more verbose name, to both make the intention explicit and to discourage usage.

It's tested via the identical deprecated implicit conversion. Is that ok? Or do you want me to switch to this one? Or do you want me to add both?

Comment (dwijnand) [reply src/main/scala/sbtbuildinfo/package.scala]: @eed3si9n wdyt?

Comment (eed3si9n) [reply src/main/scala/sbtbuildinfo/package.scala]: `outOfGraphUnsafe`?

**Pull request merged**